controllers/btpoperator_controller.go

```diff
@@ -1131,3 +1179,567 @@ func (r *BtpOperatorReconciler) watchConfigPredicates() predicate.Funcs {
+	if err != nil {
+		return false, err
+	}
+
+	webhookCertificate, err := r.getCertificateFromSecret(ctx, WebhookSecret)
+	if err != nil {
+		return false, err
+	}
+
+	ok, err := certs.VerifyIfLeafIsSignedByGivenCA(caCertificate, webhookCertificate)
+	if err != nil {
+		return false, err
+	}
+
+	return ok, nil
+}
+
+func (r *BtpOperatorReconciler) doesCertificateExpireSoon(ctx context.Context, secretName string) (bool, error) {
+	certificate, err := r.getCertificateFromSecret(ctx, secretName)
+
+	if err != nil {
+		return false, err
+	}
+	certificateDecoded, err := certs.TryDecodeCertificate(certificate)
+	if err != nil {
+		return true, err
+	}
+	certificateTemplate, err := x509.ParseCertificate(certificateDecoded.Bytes)
+	if err != nil {
+		return false, err
+	}
+
+	expirationTriggerBound := certificateTemplate.NotAfter.Add(ExpirationBoundary)
+	expiresSoon := time.Now().UTC().After(expirationTriggerBound)
+	return expiresSoon, nil
+}
+
+func (r *BtpOperatorReconciler) getDataFromSecret(ctx context.Context, name string) (map[string][]byte, error) {
+	secret := &corev1.Secret{}
+	if err := r.Get(ctx, client.ObjectKey{Namespace: ChartNamespace, Name: name}, secret); err != nil {
+		return nil, err
+	}
+	return secret.Data, nil
+}
+
+func (r *BtpOperatorReconciler) getCertificateFromSecret(ctx context.Context, secretName string) ([]byte, error) {
+	data, err := r.getDataFromSecret(ctx, secretName)
+	if err != nil {
+		return []byte{}, err
+	}
+	key, err := r.mapSecretNameToSecretDataKey(secretName)
+	if err != nil {
+		return []byte{}, err
+	}
+	cert, err := r.getValueByKey(r.buildKeyNameWithExtension(key, CertificatePostfix), data)
+	if err != nil {
+		return []byte{}, err
+	}
+	return cert, nil
+}
+
+func (r *BtpOperatorReconciler) mapSecretNameToSecretDataKey(secretName string) (string, error) {
+	switch secretName {
+	case CaSecret:
+		return CaSecretDataPrefix, nil
+	case WebhookSecret:
+		return WebhookSecretDataPrefix, nil
+	default:
+		return "", fmt.Errorf("not found secret data key for secret name: %s", secretName)
+	}
+}
+
+func (r *BtpOperatorReconciler) buildKeyNameWithExtension(filename, extension string) string {
+	return fmt.Sprintf("%s.%s", filename, extension)
+}
+
+func (r *BtpOperatorReconciler) structToByteArray(s any) ([]byte, error) {
+	var buffer bytes.Buffer
+	encoder := gob.NewEncoder(&buffer)
+	err := encoder.Encode(s)
+	if err != nil {
+		return []byte{}, err
+	}
+
+	return buffer.Bytes(), nil
+}
+
+func (r *BtpOperatorReconciler) getValueByKey(key string, data map[string][]byte) ([]byte, error) {
+	value, ok := data[key]
+	if !ok {
+		return nil, fmt.Errorf("while getting data for key: %s", key)
+	}
+	if value == nil || len(value) == 0 {
+		return nil, fmt.Errorf("empty data for key: %s", key)
+	}
+	return value, nil
+}
+
+func (r *BtpOperatorReconciler) buildSecretWithDataAndLabels(name string, data map[string][]byte, labels map[string]string) *corev1.Secret {
+	return &corev1.Secret{
+		TypeMeta: metav1.TypeMeta{
```

Comment:
Kind and APIVersion should be populated by default when you create a Secret, no need to specify TypeMeta